cloudify_aws/s3/bucket.py

```diff
@@ -0,0 +1,178 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+#    * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    * See the License for the specific language governing permissions and
+#    * limitations under the License.
+
+# Stdlib imports
+import string
+
+# Third party imports
+from botocore.exceptions import ClientError
+
+# Cloudify imports
+from cloudify import ctx
+from cloudify.decorators import operation
+from cloudify.exceptions import NonRecoverableError
+
+# This package imports
+from cloudify_aws import constants
+from cloudify_aws.base import AwsBaseNode
+from cloudify_aws.connection import EC2ConnectionClient
+
+
+@operation
+def create(args=None, **_):
+    return Bucket().created(args)
+
+
+@operation
+def delete(args=None, **_):
+    return Bucket().deleted(args)
+
+
+class Bucket(AwsBaseNode):
+    def __init__(self, client=None):
+        client = client or EC2ConnectionClient().client3('s3')
+        super(Bucket, self).__init__(
+            constants.S3_BUCKET['AWS_RESOURCE_TYPE'],
+            constants.S3_BUCKET['REQUIRED_PROPERTIES'],
+            client=client,
+            resource_id_key=constants.S3_BUCKET['RESOURCE_ID_KEY'],
+        )
+        self.not_found_error = constants.S3_BUCKET['NOT_FOUND_ERROR']
+        self.get_all_handler = {
+            'function': self.client.list_buckets,
+            'argument': 'Bucket',
+            'results_key': 'Buckets',
+            'id_key': 'Name',
+        }
+        self.resource_id = ctx.node.properties['name']
+        if self.is_external_resource:
+            ctx.instance.runtime_properties[constants.EXTERNAL_RESOURCE_ID] = (
+                ctx.node.properties['name']
+            )
+
+    def create(self, args=None):
+        self.validate_bucket_name(ctx.node.properties['name'])
+
+        self.execute(
+            self.client.create_bucket,
+            {
+                'Bucket': ctx.node.properties['name'],
+                'ACL': ctx.node.properties['permissions'],
+            }
+        )
+        ctx.instance.runtime_properties[constants.EXTERNAL_RESOURCE_ID] = (
+            ctx.node.properties['name']
+        )
+        index = ctx.node.properties['website_index_page']
+        error = ctx.node.properties['website_error_page']
+        if (index, error) == ('', ''):
+            # Neither the index nor the error page were defined, this bucket is
+            # not intended to be a website
+            pass
+        elif '' in (index, error):
+            raise NonRecoverableError(
+                'For the bucket to be configured as a website, both '
+                'website_index_page and website_error_page must be set.'
+            )
+        else:
+            if '/' in index:
+                raise NonRecoverableError(
+                    'S3 bucket website default page must not contain a /'
+                )
+            self.execute(
```

Comment:
Does it make sense to use this as the configure or start operation. I do not like having two api methods in the same operation.